### PR TITLE
[X86][AVX10.2] Support AVX10.2-CONVERT new instructions.

### PR DESCRIPTION
**Pull request by FreddyLeaf**
Ref.: https://cdrdv2.intel.com/v1/dl/getContent/828965

### CODE REVIEW

**Review by phoebewang — COMMENTED** [review]
Just found I didn't publish the comments..

File: clang/lib/Headers/avx10_2_512convertintrin.h [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+}
+
+static __inline__ __m512i __DEFAULT_FN_ATTRS512 _mm512_mask_cvtne2ph_phf8(
+    __m512i __W, __mmask64 __U, __m512h __A, __m512h __B) {
+  return (__m512i)__builtin_ia32_selectb_512(
+      (__mmask64)__U, (__v64qi)_mm512_cvtne2ph_phf8(__A, __B), (__v64qi)__W);
+}
+
+static __inline__ __m512i __DEFAULT_FN_ATTRS512
+_mm512_cvtne2ph2hf8s_phf8(__m512h __A, __m512h __B) {
```
[the remark this answers is not in the document]
Comment (FreddyLeaf) [reply clang/lib/Headers/avx10_2_512convertintrin.h]: renamed to `_mm512_cvtnes2ph_phf8`

File: clang/lib/Headers/avx10_2_512convertintrin.h [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+  return (__m512i)__builtin_ia32_vcvtne2ph2hf8s_512((__v32hf)(__A),
+                                                    (__v32hf)(__B));
+}
+
+static __inline__ __m512i __DEFAULT_FN_ATTRS512 _mm512_mask_cvtne2ph2hf8s_phf8(
```
[the remark this answers is not in the document]
Comment (FreddyLeaf) [reply clang/lib/Headers/avx10_2_512convertintrin.h]: renamed to `_mm512_mask_cvtnes2ph_phf8`

File: clang/lib/Headers/avx10_2_512convertintrin.h [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+
+static __inline__ __m512h __DEFAULT_FN_ATTRS512 _mm512_cvtx2ps_ph(__m512 __A,
+                                                                  __m512 __B) {
+  return (__m512h)__builtin_ia32_vcvt2ps2phx512_mask(
+      (__v16sf)__A, (__v16sf)__B, (__v32hf)_mm512_setzero_ph(), (__mmask32)(-1),
+      _MM_FROUND_CUR_DIRECTION);
+}
+
+static __inline__ __m512h __DEFAULT_FN_ATTRS512
+_mm512_mask_cvtx2ps_ph(__m512h __W, __mmask32 __U, __m512 __A, __m512 __B) {
```
[the remark this answers is not in the document]
Comment (FreddyLeaf) [reply clang/lib/Headers/avx10_2_512convertintrin.h:32]: good catch

File: clang/include/clang/Basic/BuiltinsX86.def [hunk 4]
```diff
@@ -2217,6 +2217,50 @@ TARGET_BUILTIN(__builtin_ia32_vcvttps2ibs512_mask, "V16UiV16fV16UiUsIi", "nV:512
...
+TARGET_BUILTIN(__builtin_ia32_vcvtne2ph2bf8s_512, "V64cV32xV32x", "nV:512:", "avx10.2-512")
+TARGET_BUILTIN(__builtin_ia32_vcvtne2ph2hf8_128, "V16cV8xV8x", "nV:128:", "avx10.2-256")
+TARGET_BUILTIN(__builtin_ia32_vcvtne2ph2hf8_256, "V32cV16xV16x", "nV:256:", "avx10.2-256")
+TARGET_BUILTIN(__builtin_ia32_vcvtne2ph2hf8_512, "V64cV32xV32x", "nV:512:", "avx10.2-512")
+TARGET_BUILTIN(__builtin_ia32_vcvtne2ph2hf8s_128, "V16cV8xV8x", "nV:128:", "avx10.2-256")
+TARGET_BUILTIN(__builtin_ia32_vcvtne2ph2hf8s_256, "V32cV16xV16x", "nV:256:", "avx10.2-256")
+TARGET_BUILTIN(__builtin_ia32_vcvtne2ph2hf8s_512, "V64cV32xV32x", "nV:512:", "avx10.2-512")
+TARGET_BUILTIN(__builtin_ia32_vcvtnehf8_2ph128_mask, "V8xV16cV8xUc", "nV:128:", "avx10.2-256")
+TARGET_BUILTIN(__builtin_ia32_vcvtnehf8_2ph256_mask, "V16xV16cV16xUs", "nV:256:", "avx10.2-256")
+TARGET_BUILTIN(__builtin_ia32_vcvtnehf8_2ph512_mask, "V32xV32cV32xUi", "nV:512:", "avx10.2-512")
```
[the remark this answers is not in the document]
Comment (FreddyLeaf) [reply clang/include/clang/Basic/BuiltinsX86.def]: Done in latest.

File: clang/lib/Headers/avx10_2_512convertintrin.h [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+_mm512_mask_cvtnehf8_ph(__m512h __W, __mmask32 __U, __m256i __A) {
+  return (__m512h)__builtin_ia32_vcvtnehf8_2ph512_mask(
+      (__v32qi)__A, (__v32hf)(__m512h)__W, (__mmask32)__U);
+}
+
+static __inline__ __m512h __DEFAULT_FN_ATTRS512
+_mm512_maskz_cvtnehf8_ph(__mmask32 __U, __m256i __A) {
+  return (__m512h)__builtin_ia32_vcvtnehf8_2ph512_mask(
+      (__v32qi)__A, (__v32hf)(__m512h)_mm512_setzero_ph(), (__mmask32)__U);
+}
```
[the remark this answers is not in the document]
Comment (FreddyLeaf) [reply clang/lib/Headers/avx10_2_512convertintrin.h]: Done in latest.

File: clang/lib/Headers/avx10_2_512convertintrin.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (FreddyLeaf) [reply clang/lib/Headers/avx10_2_512convertintrin.h]: Done in latest.

File: clang/lib/Headers/avx10_2_512convertintrin.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (FreddyLeaf) [reply clang/lib/Headers/avx10_2_512convertintrin.h]: Done in latest.

File: llvm/lib/Target/X86/X86InstrAVX10.td [hunk 6]
```diff
@@ -624,3 +624,440 @@ defm VCVTTPS2IUBS : avx10_sat_cvt_base<0x6a, "vcvttps2iubs", SchedWriteVecIMul,
...
+                               (_.VT (OpNodeRnd (_Src.VT _Src.RC:$src1),
+                                                (_Src.VT _Src.RC:$src2), (i32 timm:$rc)))>,
+                              EVEX, VVVV, EVEX_B, EVEX_RC, PD, Sched<[sched]>;
+}
+
+multiclass avx10_cvt2ps2ph<bits<8> opc, string OpcodeStr,
+                             X86SchedWriteWidths sched,
+                             AVX512VLVectorVTInfo _SrcVTInfo,
+                             AVX512VLVectorVTInfo _DstVTInfo,
+                             SDNode OpNode, SDNode OpNodeRnd> {
```
[the remark this answers is not in the document]
Comment (FreddyLeaf) [reply llvm/lib/Target/X86/X86InstrAVX10.td]: Done in latest.

File: llvm/lib/Target/X86/X86InstrAVX10.td [hunk 7]
```diff
@@ -624,3 +624,440 @@ defm VCVTTPS2IUBS : avx10_sat_cvt_base<0x6a, "vcvttps2iubs", SchedWriteVecIMul,
...
+           X86VectorVTInfo vt_dst, X86VectorVTInfo vt_src1,
+           X86VectorVTInfo vt_src2, SDPatternOperator OpNode,
+           SDPatternOperator MaskOpNode, X86FoldableSchedWrite sched,
+           string Broadcast = vt_src2.BroadcastStr,
+           X86MemOperand MemOp = vt_src2.MemOp,
+           RegisterClass MaskRC = vt_src2.KRCWM,
+           dag LdDAG = (vt_dst.VT (OpNode (vt_src1.VT vt_src1.RC:$src1),
+                                  (vt_src2.VT (vt_src2.LdFrag addr:$src2)))),
+           dag MaskLdDAG = (vt_dst.VT (MaskOpNode (vt_src1.VT vt_src1.RC:$src1),
+                                      (vt_src2.VT (vt_src2.LdFrag addr:$src2))))> {
```
[the remark this answers is not in the document]
Comment (FreddyLeaf) [reply llvm/lib/Target/X86/X86InstrAVX10.td]: Done in latest.

File: llvm/lib/Target/X86/X86InstrAVX10.td [hunk 8]
```diff
@@ -624,3 +624,440 @@ defm VCVTTPS2IUBS : avx10_sat_cvt_base<0x6a, "vcvttps2iubs", SchedWriteVecIMul,
                                       avx512vl_i32_info, avx512vl_f32_info,
                                       X86vcvttp2iubsSAE>,
                     AVX512PDIi8Base, T_MAP5, EVEX_CD8<32, CD8VF>;
+
+//-------------------------------------------------
+// AVX10 CONVERT instructions
+//-------------------------------------------------
+
+multiclass avx10_cvt2ps2ph_rc<bits<8> opc, string OpcodeStr, X86FoldableSchedWrite sched,
```
[the remark this answers is not in the document]
Comment (FreddyLeaf) [reply llvm/lib/Target/X86/X86InstrAVX10.td:632]: Done in latest.

File: llvm/lib/Target/X86/X86InstrAVX10.td [hunk 9]
```diff
@@ -624,3 +624,328 @@ defm VCVTTPS2IUBS : avx10_sat_cvt_base<0x6a, "vcvttps2iubs", SchedWriteVecIMul,
...
+              (!cast<Instruction>(NAME # "Z128rmbkz") maskRC128:$mask, addr:$src)>;
+  }
+}
+
+defm VCVTNEPH2BF8 : avx512_cvt_trunc_ne<0x74, "vcvtneph2bf8", avx512vl_i8_info,
+                                        avx512vl_f16_info, SchedWriteCvtPD2PS,
+                                        X86vcvtneph2bf8, [HasAVX10_2], [HasAVX10_2_512]>,
+                                        T8, XS, EVEX_CD8<16, CD8VF>;
+defm VCVTNEPH2BF8 : avx10_convert_2op_pat<avx512vl_i8_info, avx512vl_f16_info,
+                                          X86vcvtneph2bf8, X86vmcvtneph2bf8>;
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86InstrAVX10.td:894]: Wrap `avx512_cvt_trunc_ne` and `avx10_convert_2op_pat` to a new class?

File: llvm/lib/Target/X86/X86InstrFragmentsSIMD.td [hunk 10]
```diff
@@ -847,6 +847,91 @@ def X86vcvttp2iubs : SDNode<"X86ISD::CVTTP2IUBS",  SDTFloatToInt>;
...
+  SDTCVecEltisVT<0, i8>, SDTCVecEltisVT<1, f16>, SDTCisSameAs<1, 2>
+]>;
+
+def SDTAVX10CONVERT_F16I8 : SDTypeProfile<1, 1, [
+  SDTCVecEltisVT<0, f16>, SDTCVecEltisVT<1, i8>
+]>;
+
+def SDTAVX10CONVERT_I8F16 : SDTypeProfile<1, 1, [
+  SDTCVecEltisVT<0, i8>, SDTCVecEltisVT<1, f16>
+]>;
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86InstrFragmentsSIMD.td:860]: Check `SDTCVecEltisVT<1, ..`

File: llvm/lib/Target/X86/X86InstrFragmentsSIMD.td [hunk 11]
```diff
@@ -847,6 +847,91 @@ def X86vcvttp2iubs : SDNode<"X86ISD::CVTTP2IUBS",  SDTFloatToInt>;
...
+def SDTAVX10CONVERT_2I8F16_MASK : SDTypeProfile<1, 4, [
+  SDTCVecEltisVT<0, i8>, SDTCisSameAs<0, 1>,
+  SDTCVecEltisVT<2, f16>, SDTCisSameAs<0, 3>, SDTCVecEltisVT<4, i1>,
+  SDTCisSameNumEltsAs<2, 4>
+]>;
+
+def X86vcvt2ps2phx : SDNode<"X86ISD::VCVT2PS2PHX",
+                     SDTypeProfile<1, 2, [SDTCVecEltisVT<0, f16>,
+                                           SDTCVecEltisVT<1, f32>,
+                                           SDTCisSameAs<1,2>]>>;
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86InstrFragmentsSIMD.td:881]: indentation

File: llvm/lib/Target/X86/X86InstrFragmentsSIMD.td [hunk 12]
```diff
@@ -847,6 +847,91 @@ def X86vcvttp2iubs : SDNode<"X86ISD::CVTTP2IUBS",  SDTFloatToInt>;
...
+def X86vcvt2ps2phxRnd : SDNode<"X86ISD::VCVT2PS2PHX_RND",
+                        SDTypeProfile<1, 3, [SDTCVecEltisVT<0, f16>,
+                                            SDTCVecEltisVT<1, f32>,
+                                            SDTCisSameAs<1,2>,
+                                            SDTCisVT<3, i32>]>>;
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86InstrFragmentsSIMD.td:886]: indentation

File: llvm/lib/Target/X86/X86InstrFragmentsSIMD.td [hunk 13]
```diff
@@ -847,6 +847,91 @@ def X86vcvttp2iubs : SDNode<"X86ISD::CVTTP2IUBS",  SDTFloatToInt>;
...
+// 3op
+def X86vcvtne2ph2bf8 : SDNode<"X86ISD::VCVTNE2PH2BF8",
+                       SDTAVX10CONVERT_I82F16>;
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86InstrFragmentsSIMD.td:889]: Align within `<`

File: llvm/lib/Target/X86/X86InstrFragmentsSIMD.td (hunk 11 above)
Comment (phoebewang) [reply llvm/lib/Target/X86/X86InstrFragmentsSIMD.td:881]: Maybe merge with `X86cvtne2ps2bf16`?

File: llvm/lib/TargetParser/X86TargetParser.cpp [hunk 14]
```diff
@@ -617,7 +617,7 @@ constexpr FeatureBitset ImpliedFeaturesAVX10_1 =
     FeatureAVX512CD | FeatureAVX512VBMI | FeatureAVX512IFMA |
     FeatureAVX512VNNI | FeatureAVX512BF16 | FeatureAVX512VPOPCNTDQ |
     FeatureAVX512VBMI2 | FeatureAVX512BITALG | FeatureVAES | FeatureVPCLMULQDQ |
-    FeatureAVX512FP16;
+    FeatureAVX512FP16 | FeatureAVX2;
```
Comment (phoebewang) [inline llvm/lib/TargetParser/X86TargetParser.cpp:620]: Why adding it? AVX512 features already imply it.

**Review by KanRobert — APPROVED** [review]
MC and foldtable change LGTM

File: llvm/include/llvm/IR/IntrinsicsX86.td [hunk 15]
```diff
@@ -7089,3 +7089,133 @@ def int_x86_avx10_mask_vcvttps2iubs512 : ClangBuiltin<"__builtin_ia32_vcvttps2iu
         DefaultAttrsIntrinsic<[llvm_v16i32_ty], [llvm_v16f32_ty, llvm_v16i32_ty, llvm_i16_ty, llvm_i32_ty],
                   [IntrNoMem, ImmArg<ArgIndex<3>>]>;
 }
+
+//===----------------------------------------------------------------------===//
+let TargetPrefix = "x86" in {
+def int_x86_avx10_mask_vcvt2ps2phx_128 : ClangBuiltin<"__builtin_ia32_vcvt2ps2phx128_mask">,
+        DefaultAttrsIntrinsic<[llvm_v8f16_ty], [llvm_v4f32_ty, llvm_v4f32_ty, llvm_v8f16_ty, llvm_i8_ty],
+                  [IntrNoMem]>;
```
[the remark this answers is not in the document]
Comment (FreddyLeaf) [reply llvm/include/llvm/IR/IntrinsicsX86.td]: [4a5a207](https://github.com/llvm/llvm-project/pull/101600/commits/4a5a2077bf72fe6d13dbf3bf4226817db7aa808a)

File: llvm/lib/Target/X86/X86InstrAVX10.td (hunk 9 above)
Comment (FreddyLeaf) [reply llvm/lib/Target/X86/X86InstrAVX10.td]: [4a5a207](https://github.com/llvm/llvm-project/pull/101600/commits/4a5a2077bf72fe6d13dbf3bf4226817db7aa808a)

File: llvm/lib/Target/X86/X86InstrFragmentsSIMD.td (hunk 10 above)
Comment (FreddyLeaf) [reply llvm/lib/Target/X86/X86InstrFragmentsSIMD.td:860]: [4a5a207](https://github.com/llvm/llvm-project/pull/101600/commits/4a5a2077bf72fe6d13dbf3bf4226817db7aa808a)

File: llvm/lib/Target/X86/X86InstrFragmentsSIMD.td (hunk 11 above)
Comment (FreddyLeaf) [reply llvm/lib/Target/X86/X86InstrFragmentsSIMD.td]: [4a5a207](https://github.com/llvm/llvm-project/pull/101600/commits/4a5a2077bf72fe6d13dbf3bf4226817db7aa808a)

File: llvm/lib/Target/X86/X86InstrFragmentsSIMD.td (hunk 12 above)
Comment (FreddyLeaf) [reply llvm/lib/Target/X86/X86InstrFragmentsSIMD.td]: [4a5a207](https://github.com/llvm/llvm-project/pull/101600/commits/4a5a2077bf72fe6d13dbf3bf4226817db7aa808a)

File: llvm/lib/Target/X86/X86InstrFragmentsSIMD.td (hunk 13 above)
Comment (FreddyLeaf) [reply llvm/lib/Target/X86/X86InstrFragmentsSIMD.td]: [4a5a207](https://github.com/llvm/llvm-project/pull/101600/commits/4a5a2077bf72fe6d13dbf3bf4226817db7aa808a)

File: llvm/lib/TargetParser/X86TargetParser.cpp (hunk 14 above)
Comment (FreddyLeaf) [reply llvm/lib/TargetParser/X86TargetParser.cpp]: [4a5a207](https://github.com/llvm/llvm-project/pull/101600/commits/4a5a2077bf72fe6d13dbf3bf4226817db7aa808a)

**Review by phoebewang — APPROVED** [review]
LGTM.

**Comment by jhuber6** [comment]
```llvm/llvm-project/build/lib/clang/20/include/avx10_2convertintrin.h:29:19: error: use of undeclared identifier '__builtin_ia32_vcvt2ps2phx128_mask'
   29 |   return (__m128h)__builtin_ia32_vcvt2ps2phx128_mask(
      |                   ^
```
I'm getting a lot of errors like this since this patch landed.

**Comment by jhuber6** [comment]
My guess is that you added these as a target dependent builtin, but made the headers always reference them.

**Pull request merged**